src/xdp-utils.c

```diff
@@ -144,11 +149,62 @@ xdp_app_info_new (XdpAppInfoKind kind)
   return app_info;
 }
 
+void
+get_app_id_from_pid (pid_t pid, XdpAppInfo *app_info)
+{
+#ifdef HAVE_LIBSYSTEMD
+  const char *app_id;
+  g_autoptr(GRegex) regex = NULL;
+  g_autoptr(GMatchInfo) match = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree char *unit = NULL;
+  int res;
+
+  res = sd_pid_get_user_unit (pid, &unit);
+  /*
+   * The session might not be managed by systemd or there could be an error
+   * fetching our own systemd units or the unit might not be started by the
+   * desktop environment (e.g. it's a script run from terminal).
+   */
+  if (res == -ENODATA || res < 0 || !unit || !g_str_has_prefix (unit, "app-"))
+    {
+      app_info->id = g_strdup ("");
+      return;
+    }
+
+  /*
+   * From https://systemd.io/DESKTOP_ENVIRONMENTS/ the format is one of:
+   * app[-<launcher>]-<ApplicationID>-<RANDOM>.scope
+   * app[-<launcher>]-<ApplicationID>-autostart.service -> no longer true since systemd v248
+   * app[-<launcher>]-<ApplicationID>[@<RANDOM>].service
+   */
+  regex = g_regex_new ("^app-(?:[[:alnum:]]+\\-)?(.+?\\..+?\\..+?)"
```

Comment:
All the string manipulation code should be split and unit tests added.

Reply:
You're right, added a unit test